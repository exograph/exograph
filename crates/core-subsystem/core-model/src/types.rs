// Copyright Exograph, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file at the root of this repository.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

//! Common support for representing GraphQL types such as `Int`, `List<Int>`, `Optional<Int>`, `Optional<List<Int>>`, etc.
//!

use async_graphql_parser::types::{BaseType, Type};
use async_graphql_value::Name;
use serde::{Deserialize, Serialize};

use crate::mapped_arena::{SerializableSlab, SerializableSlabIndex};

/// A type that can be used as a type for fields and return types
/// Currently supports only list and optional decorations
#[derive(Serialize, Deserialize, Debug, Clone)]
pub enum FieldType<T> {
    Plain(T),
    List(Box<FieldType<T>>),
    Optional(Box<FieldType<T>>),
}

pub trait Named {
    fn name(&self) -> &str;
}

impl<T> FieldType<T> {
    /// Return the inner type, if any
    pub fn inner(&self) -> Option<&FieldType<T>> {
        match self {
            FieldType::Plain(_) => None,
            FieldType::List(inner) | FieldType::Optional(inner) => Some(inner),
        }
    }

    /// Return the base type (i.e. by removing optional decoration)
    pub fn base_type(&self) -> &FieldType<T> {
        match self {
            FieldType::Optional(inner_typ) => inner_typ.as_ref(),
            _ => self,
        }
    }

    /// Return the innermost (i.e. leaf) type
    pub fn innermost(&self) -> &T {
        match self {
            FieldType::Plain(inner) => inner,
            FieldType::List(inner) | FieldType::Optional(inner) => inner.innermost(),
        }
    }

    /// Wrap the destination type to match the structure of the source type
    pub fn wrap<D>(&self, dest: D) -> FieldType<D> {
        match self {
            FieldType::Plain(_) => FieldType::Plain(dest),
            FieldType::Optional(optional) => FieldType::Optional(Box::new(optional.wrap(dest))),
            FieldType::List(list) => FieldType::List(Box::new(list.wrap(dest))),
        }
    }
}

impl<T: Clone> FieldType<T> {
    /// Compute the optional version of the given type
    pub fn optional(&self) -> Self {
        match self {
            FieldType::Optional(_) => self.clone(), // Already optional
            _ => FieldType::Optional(Box::new(self.clone())),
        }
    }
}

/// Transforms the type into an introspection type
///
/// The complexity of this function is due to the fact that the GraphQL spec and hence the
/// introspection type ([`Type`]) does not support nested optionals. However, [`FieldType`]
/// being more general, does. This function will panic if it encounters a nested optional.
impl<T: Named> From<&FieldType<T>> for Type {
    fn from(ft: &FieldType<T>) -> Self {
        match ft {
            FieldType::Plain(base) => Type {
                base: BaseType::Named(Name::new(base.name())),
                nullable: false,
            },
            FieldType::Optional(underlying) => {
                let Type { base, nullable } = underlying.as_ref().into();

                if nullable {
                    panic!("Optional type cannot be nested")
                }
                Type {
                    base,
                    nullable: true,
                }
            }
            FieldType::List(underlying) => Type {
                base: BaseType::List(Box::new(underlying.as_ref().into())),
                nullable: false,
            },
        }
    }
}

impl<T: Named> Named for FieldType<T> {
    fn name(&self) -> &str {
        match self {
            FieldType::Plain(plain) => plain.name(),
            FieldType::List(list) => list.name(),
            FieldType::Optional(optional) => optional.name(),
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct BaseOperationReturnType<T> {
    pub associated_type_id: SerializableSlabIndex<T>,
    pub type_name: String,
}

impl<T> Named for BaseOperationReturnType<T> {
    fn name(&self) -> &str {
        &self.type_name
    }
}

pub type OperationReturnType<T> = FieldType<BaseOperationReturnType<T>>;

impl<T> OperationReturnType<T> {
    pub fn typ<'a>(&'a self, types: &'a SerializableSlab<T>) -> &T {
        match self {
            OperationReturnType::Plain(BaseOperationReturnType {
                associated_type_id, ..
            }) => &types[*associated_type_id],
            OperationReturnType::List(underlying) | OperationReturnType::Optional(underlying) => {
                underlying.typ(types)
            }
        }
    }

    pub fn type_name(&self) -> &str {
        match self {
            OperationReturnType::Plain(BaseOperationReturnType { type_name, .. }) => type_name,
            OperationReturnType::List(underlying) | OperationReturnType::Optional(underlying) => {
                underlying.type_name()
            }
        }
    }
}
