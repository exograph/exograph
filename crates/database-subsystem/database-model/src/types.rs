use super::access::Access;
use super::{column_id::ColumnId, relation::DatabaseRelation};
use crate::model::ModelDatabaseSystem;
use crate::operation::{DatabaseQuery, DatabaseQueryParameter};
use async_graphql_parser::types::{
    BaseType, FieldDefinition, InputObjectType, InputValueDefinition, ObjectType, Type,
    TypeDefinition, TypeKind,
};
use async_graphql_value::Name;
use core_model::mapped_arena::{SerializableSlab, SerializableSlabIndex};

use core_model::type_normalization::{
    default_positioned, default_positioned_name, FieldDefinitionProvider, InputValueProvider,
    TypeDefinitionProvider, TypeModifier,
};
use payas_sql::PhysicalTable;

use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct DatabaseType {
    pub name: String,
    pub plural_name: String,
    pub kind: DatabaseTypeKind,
    pub is_input: bool, // Is this to be used as an input field (such as an argument in a mutation)? Needed for introspection
}

impl DatabaseType {
    pub fn model_fields(&self) -> Vec<&DatabaseField> {
        match &self.kind {
            DatabaseTypeKind::Primitive => vec![],
            DatabaseTypeKind::Composite(DatabaseCompositeType { fields, .. }) => {
                fields.iter().collect()
            }
        }
    }

    pub fn model_field(&self, name: &str) -> Option<&DatabaseField> {
        self.model_fields()
            .into_iter()
            .find(|model_field| model_field.name == name)
    }

    pub fn pk_field(&self) -> Option<&DatabaseField> {
        match &self.kind {
            DatabaseTypeKind::Primitive => None,
            DatabaseTypeKind::Composite(composite_type) => composite_type.pk_field(),
        }
    }

    pub fn pk_column_id(&self) -> Option<ColumnId> {
        match &self.kind {
            DatabaseTypeKind::Primitive => None,
            DatabaseTypeKind::Composite(composite_type) => composite_type.pk_column_id(),
        }
    }

    pub fn table_id(&self) -> Option<SerializableSlabIndex<PhysicalTable>> {
        match &self.kind {
            DatabaseTypeKind::Composite(DatabaseCompositeType { table_id, .. }) => Some(*table_id),
            _ => None,
        }
    }

    pub fn is_primitive(&self) -> bool {
        matches!(&self.kind, DatabaseTypeKind::Primitive)
    }
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[allow(clippy::large_enum_variant)]
pub enum DatabaseTypeKind {
    Primitive,
    Composite(DatabaseCompositeType),
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct DatabaseCompositeType {
    pub fields: Vec<DatabaseField>,
    pub table_id: SerializableSlabIndex<PhysicalTable>,
    pub pk_query: SerializableSlabIndex<DatabaseQuery>,
    pub collection_query: SerializableSlabIndex<DatabaseQuery>,
    pub access: Access,
}

impl DatabaseCompositeType {
    pub fn get_field_by_name(&self, name: &str) -> Option<&DatabaseField> {
        self.fields.iter().find(|field| field.name == name)
    }

    pub fn pk_field(&self) -> Option<&DatabaseField> {
        self.fields
            .iter()
            .find(|field| matches!(&field.relation, DatabaseRelation::Pk { .. }))
    }

    pub fn pk_column_id(&self) -> Option<ColumnId> {
        self.pk_field()
            .and_then(|pk_field| pk_field.relation.self_column())
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub enum DatabaseTypeModifier {
    Optional,
    NonNull,
    List,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct DatabaseField {
    pub name: String,
    pub typ: DatabaseFieldType,
    pub relation: DatabaseRelation,
    pub has_default_value: bool, // does this field have a default value?
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub enum DatabaseFieldType {
    Optional(Box<DatabaseFieldType>),
    Reference {
        type_id: SerializableSlabIndex<DatabaseType>,
        is_primitive: bool, // A way to know which arena to look up the type in
        type_name: String,
    },
    List(Box<DatabaseFieldType>),
}

impl DatabaseFieldType {
    pub fn type_id(&self) -> &SerializableSlabIndex<DatabaseType> {
        match self {
            DatabaseFieldType::Optional(underlying) | DatabaseFieldType::List(underlying) => {
                underlying.type_id()
            }
            DatabaseFieldType::Reference { type_id, .. } => type_id,
        }
    }

    pub fn is_primitive(&self) -> bool {
        match self {
            DatabaseFieldType::Optional(underlying) | DatabaseFieldType::List(underlying) => {
                underlying.is_primitive()
            }
            DatabaseFieldType::Reference { is_primitive, .. } => *is_primitive,
        }
    }

    pub fn base_type<'a>(&self, types: &'a SerializableSlab<DatabaseType>) -> &'a DatabaseType {
        match self {
            DatabaseFieldType::Optional(underlying) | DatabaseFieldType::List(underlying) => {
                underlying.base_type(types)
            }
            DatabaseFieldType::Reference { type_id, .. } => &types[*type_id],
        }
    }

    pub fn type_name(&self) -> &str {
        match self {
            DatabaseFieldType::Optional(underlying) | DatabaseFieldType::List(underlying) => {
                underlying.type_name()
            }
            DatabaseFieldType::Reference { type_name, .. } => type_name,
        }
    }

    pub fn optional(&self) -> Self {
        match self {
            DatabaseFieldType::Optional(_) => self.clone(),
            _ => DatabaseFieldType::Optional(Box::new(self.clone())),
        }
    }
}

impl TypeDefinitionProvider<ModelDatabaseSystem> for DatabaseType {
    fn type_definition(&self, system: &ModelDatabaseSystem) -> TypeDefinition {
        match &self.kind {
            DatabaseTypeKind::Primitive => TypeDefinition {
                extend: false,
                description: None,
                name: default_positioned_name(&self.name),
                directives: vec![],
                kind: TypeKind::Scalar,
            },
            DatabaseTypeKind::Composite(DatabaseCompositeType {
                fields: model_fields,
                ..
            }) => {
                let kind = if self.is_input {
                    let fields = model_fields
                        .iter()
                        .map(|model_field| default_positioned(model_field.input_value()))
                        .collect();
                    TypeKind::InputObject(InputObjectType { fields })
                } else {
                    let fields: Vec<_> = model_fields
                        .iter()
                        .map(|model_field| default_positioned(model_field.field_definition(system)))
                        .collect();

                    TypeKind::Object(ObjectType {
                        implements: vec![],
                        fields,
                    })
                };
                TypeDefinition {
                    extend: false,
                    description: None,
                    name: default_positioned_name(&self.name),
                    directives: vec![],
                    kind,
                }
            }
        }
    }
}

impl FieldDefinitionProvider<ModelDatabaseSystem> for DatabaseField {
    fn field_definition(&self, system: &ModelDatabaseSystem) -> FieldDefinition {
        let field_type = default_positioned(compute_type(&self.typ));

        let arguments = match self.relation {
            DatabaseRelation::Pk { .. }
            | DatabaseRelation::Scalar { .. }
            | DatabaseRelation::ManyToOne { .. } => {
                vec![]
            }
            DatabaseRelation::OneToMany { other_type_id, .. } => {
                let other_type = &system.database_types[other_type_id];
                match &other_type.kind {
                    DatabaseTypeKind::Primitive => panic!(),
                    DatabaseTypeKind::Composite(kind) => {
                        let collection_query = kind.collection_query;
                        let collection_query = &system.queries[collection_query];

                        let DatabaseQueryParameter {
                            predicate_param,
                            order_by_param,
                            limit_param,
                            offset_param,
                        } = &collection_query.parameter;

                        let predicate_parameter_arg =
                            predicate_param.as_ref().map(|p| p.input_value());
                        let order_by_parameter_arg =
                            order_by_param.as_ref().map(|p| p.input_value());
                        let limit_arg = limit_param.as_ref().map(|p| p.input_value());
                        let offset_arg = offset_param.as_ref().map(|p| p.input_value());

                        vec![
                            predicate_parameter_arg,
                            order_by_parameter_arg,
                            limit_arg,
                            offset_arg,
                        ]
                        .into_iter()
                        .flatten()
                        .map(default_positioned)
                        .collect()
                    }
                }
            }
        };

        FieldDefinition {
            description: None,
            name: default_positioned_name(&self.name),
            arguments,
            ty: field_type,
            directives: vec![],
        }
    }
}

pub fn compute_type(typ: &DatabaseFieldType) -> Type {
    fn compute_base_type(typ: &DatabaseFieldType) -> BaseType {
        match typ {
            DatabaseFieldType::Optional(underlying) => compute_base_type(underlying),
            DatabaseFieldType::Reference { type_name, .. } => BaseType::Named(Name::new(type_name)),
            DatabaseFieldType::List(underlying) => {
                BaseType::List(Box::new(compute_type(underlying)))
            }
        }
    }

    match typ {
        DatabaseFieldType::Optional(underlying) => Type {
            base: compute_base_type(underlying),
            nullable: true,
        },
        DatabaseFieldType::Reference { type_name, .. } => Type {
            base: BaseType::Named(Name::new(type_name)),
            nullable: false,
        },
        DatabaseFieldType::List(underlying) => Type {
            base: BaseType::List(Box::new(compute_type(underlying))),
            nullable: false,
        },
    }
}

impl From<&DatabaseTypeModifier> for TypeModifier {
    fn from(modifier: &DatabaseTypeModifier) -> Self {
        match modifier {
            DatabaseTypeModifier::Optional => TypeModifier::Optional,
            DatabaseTypeModifier::NonNull => TypeModifier::NonNull,
            DatabaseTypeModifier::List => TypeModifier::List,
        }
    }
}

// We need to a special case for the GqlField type, so that we can properly
// created nested types such as Optional(List(List(String))). The blanket impl
// above will not work for nested types like these.
impl InputValueProvider for DatabaseField {
    fn input_value(&self) -> InputValueDefinition {
        let field_type = default_positioned(compute_type(&self.typ));

        InputValueDefinition {
            description: None,
            name: default_positioned_name(&self.name),
            ty: field_type,
            default_value: None,
            directives: vec![],
        }
    }
}
