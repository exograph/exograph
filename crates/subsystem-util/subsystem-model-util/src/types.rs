use async_graphql_parser::types::{
    FieldDefinition, InputObjectType, InputValueDefinition, ObjectType, TypeDefinition, TypeKind,
};

use core_model::{
    mapped_arena::{SerializableSlab, SerializableSlabIndex},
    type_normalization::{
        default_positioned, default_positioned_name, FieldDefinitionProvider, InputValueProvider,
        TypeDefinitionProvider, TypeModifier,
    },
    types::{FieldType, Named},
};

use serde::{Deserialize, Serialize};

use crate::access::Access;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ServiceType {
    pub name: String,
    pub kind: ServiceTypeKind,
    pub is_input: bool, // Is this to be used as an input field (such as an argument in a mutation)? Needed for introspection
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[allow(clippy::large_enum_variant)]
pub enum ServiceTypeKind {
    Primitive,
    Composite(ServiceCompositeType),
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ServiceCompositeType {
    pub fields: Vec<ServiceField>,
    pub is_input: bool,
    pub access: Access,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub enum ServiceTypeModifier {
    Optional,
    NonNull,
    List,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ServiceField {
    pub name: String,
    pub typ: FieldType<ServiceFieldType>,
    pub has_default_value: bool, // does this field have a default value?
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ServiceFieldType {
    pub type_id: SerializableSlabIndex<ServiceType>,
    pub type_name: String,
}

impl Named for ServiceFieldType {
    fn name(&self) -> &str {
        &self.type_name
    }
}

impl From<&ServiceTypeModifier> for TypeModifier {
    fn from(modifier: &ServiceTypeModifier) -> Self {
        match modifier {
            ServiceTypeModifier::Optional => TypeModifier::Optional,
            ServiceTypeModifier::NonNull => TypeModifier::NonNull,
            ServiceTypeModifier::List => TypeModifier::List,
        }
    }
}

impl TypeDefinitionProvider<SerializableSlab<ServiceType>> for ServiceType {
    fn type_definition(&self, service_types: &SerializableSlab<ServiceType>) -> TypeDefinition {
        match &self.kind {
            ServiceTypeKind::Primitive => TypeDefinition {
                extend: false,
                description: None,
                name: default_positioned_name(&self.name),
                directives: vec![],
                kind: TypeKind::Scalar,
            },
            ServiceTypeKind::Composite(ServiceCompositeType {
                fields: model_fields,
                ..
            }) => {
                let kind = if self.is_input {
                    let fields = model_fields
                        .iter()
                        .map(|model_field| default_positioned(model_field.input_value()))
                        .collect();
                    TypeKind::InputObject(InputObjectType { fields })
                } else {
                    let fields: Vec<_> = model_fields
                        .iter()
                        .map(|model_field| {
                            default_positioned(model_field.field_definition(service_types))
                        })
                        .collect();

                    TypeKind::Object(ObjectType {
                        implements: vec![],
                        fields,
                    })
                };
                TypeDefinition {
                    extend: false,
                    description: None,
                    name: default_positioned_name(&self.name),
                    directives: vec![],
                    kind,
                }
            }
        }
    }
}

impl FieldDefinitionProvider<SerializableSlab<ServiceType>> for ServiceField {
    fn field_definition(&self, _service_types: &SerializableSlab<ServiceType>) -> FieldDefinition {
        let field_type = default_positioned(self.typ.to_introspection_type());

        FieldDefinition {
            description: None,
            name: default_positioned_name(&self.name),
            arguments: vec![],
            ty: field_type,
            directives: vec![],
        }
    }
}

// We need to a special case for the GqlField type, so that we can properly
// created nested types such as Optional(List(List(String))). The blanket impl
// above will not work for nested types like these.
impl InputValueProvider for ServiceField {
    fn input_value(&self) -> InputValueDefinition {
        let field_type = default_positioned(self.typ.to_introspection_type());

        InputValueDefinition {
            description: None,
            name: default_positioned_name(&self.name),
            ty: field_type,
            default_value: None,
            directives: vec![],
        }
    }
}
