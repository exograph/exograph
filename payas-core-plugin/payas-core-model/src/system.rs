use std::collections::HashMap;

use serde::{Deserialize, Serialize};

/// A type to represent the index of an interceptor within a subsystem.
///
/// This (instead of a simple `usize`) is used to make it intentional that the index is not
/// used for anything else than indexing into the interceptor list.
#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct InterceptorIndex(pub usize);

/// A type to represent the index of an interceptor across subsystems.
#[derive(Serialize, Deserialize, Debug)]
pub struct InterceptorIndexWithSubsystemIndex {
    pub subsystem_index: usize,
    pub interceptor_index: InterceptorIndex,
}

impl InterceptorIndexWithSubsystemIndex {
    pub fn new(subsystem_index: usize, interceptor_index: InterceptorIndex) -> Self {
        Self {
            subsystem_index,
            interceptor_index,
        }
    }
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Subsystem {
    pub id: String,
    pub subsystem_index: usize,
    pub serialized_subsystem: Vec<u8>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct System {
    pub subsystems: Vec<Subsystem>,
    pub query_interception_map: InterceptionMap,
    pub mutation_interception_map: InterceptionMap,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct InterceptionMap {
    pub map: HashMap<String, Vec<InterceptorIndexWithSubsystemIndex>>,
}

impl InterceptionMap {
    pub fn get(&self, operation_name: &str) -> Option<&Vec<InterceptorIndexWithSubsystemIndex>> {
        self.map.get(operation_name)
    }
}
