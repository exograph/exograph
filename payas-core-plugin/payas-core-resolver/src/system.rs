use async_graphql_parser::{
    types::{FieldDefinition, TypeDefinition},
    Positioned,
};
use payas_core_model::serializable_system::InterceptionMap;
use thiserror::Error;
use tracing::instrument;

use crate::{
    plugin::SubsystemResolver, request_context::RequestContext,
    validation::validation_error::ValidationError, OperationsPayload, QueryResponse,
};

pub struct SystemResolver {
    pub subsystem_resolvers: Vec<Box<dyn SubsystemResolver + Send + Sync>>,
    pub query_interception_map: InterceptionMap,
    pub mutation_interception_map: InterceptionMap,

    pub allow_introspection: bool,
}

impl SystemResolver {
    /// Resolve the provided top-level operation.
    ///
    /// Goes through the FieldResolver for ValidatedOperation (and thus get free support for `resolve_fields`)
    /// so that we can support fragments in top-level queries in such as:
    ///
    /// ```graphql
    /// {
    ///   ...query_info
    /// }
    ///
    /// fragment query_info on Query {
    ///   __type(name: "Concert") {
    ///     name
    ///   }
    ///
    ///   __schema {
    ///       types {
    ///       name
    ///     }
    ///   }
    /// }
    /// ```
    #[instrument(
        name = "OperationsExecutor::resolve"
        skip_all
        )]
    pub async fn resolve<'e>(
        &'e self,
        operations_payload: OperationsPayload,
        request_context: &'e RequestContext<'e>,
    ) -> Result<Vec<(String, QueryResponse)>, ExecutionError> {
        // let operation = self.validate_operation(operations_payload)?;

        todo!()
    }

    pub fn schema_queries(&self) -> Vec<Positioned<FieldDefinition>> {
        self.subsystem_resolvers
            .iter()
            .fold(vec![], |mut acc, resolver| {
                acc.extend(resolver.schema_queries());
                acc
            })
    }

    pub fn schema_mutations(&self) -> Vec<Positioned<FieldDefinition>> {
        self.subsystem_resolvers
            .iter()
            .fold(vec![], |mut acc, resolver| {
                acc.extend(resolver.schema_mutations());
                acc
            })
    }

    pub fn schema_types(&self) -> Vec<TypeDefinition> {
        self.subsystem_resolvers
            .iter()
            .fold(vec![], |mut acc, resolver| {
                acc.extend(resolver.schema_types());
                acc
            })
    }
}

// Temporary
#[derive(Debug, Error)]
pub enum ExecutionError {
    #[error("Execution error: {0}")]
    Generic(String),

    #[error("{0}")]
    Validation(#[from] ValidationError),
}

impl ExecutionError {
    // Message that should be emitted when the error is returned to the user.
    // This should hide any internal details of the error.
    // TODO: Log the details of the error.
    pub fn user_error_message(&self) -> String {
        "todo".to_string()
    }
}
