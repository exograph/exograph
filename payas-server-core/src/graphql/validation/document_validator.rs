use async_graphql_parser::types::{DocumentOperations, ExecutableDocument};
use async_graphql_value::Name;
use payas_model::model::system::ModelSystem;
use serde_json::{Map, Value};
use tracing::instrument;

use crate::graphql::{
    introspection::definition::schema::Schema, validation::validation_error::ValidationError,
};

use super::{operation::ValidatedOperation, operation_validator::OperationValidator};

/// Context for validating a document.
pub struct DocumentValidator<'a> {
    system: &'a ModelSystem,
    schema: &'a Schema,
    operation_name: Option<String>,
    variables: Option<Map<String, Value>>,
}

impl<'a> DocumentValidator<'a> {
    pub fn new(
        system: &'a ModelSystem,
        schema: &'a Schema,
        operation_name: Option<String>,
        variables: Option<Map<String, Value>>,
    ) -> Self {
        Self {
            system,
            schema,
            operation_name,
            variables,
        }
    }

    /// Validate the query payload.
    ///
    /// Validations performed:
    /// - Validate that either there is only one operation or the operation name specified matches one of the operations in the document
    /// - Validate that there is at least one operation
    /// - Other validations are delegated to the operation validator
    #[instrument(
        name = "DocumentValidator::validate"
        skip(self, document)
        )]
    pub fn validate(
        self,
        document: ExecutableDocument,
    ) -> Result<ValidatedOperation, ValidationError> {
        let (operation_name, raw_operation) = match document.operations {
            DocumentOperations::Single(operation) => Ok((self.operation_name, operation)),
            DocumentOperations::Multiple(mut operations) => {
                if operations.is_empty() {
                    Err(ValidationError::NoOperationFound)
                } else {
                    match self.operation_name {
                        None if operations.len() == 1 => {
                            // Per https://graphql.org/learn/queries/#operation-name, `operationName` is required
                            // only for multiple operations, but async-graphql parses a named operation (`query Foo { ... }`)
                            // to `DocumentOperations::Multiple` even if there is only one operation. So we add an additional
                            // check here to make sure that the operation name is enforced only for truly multiple operations.

                            // This unwrap is okay because we already check that there is exactly one operation.
                            let (operation_name, operation) =
                                operations.into_iter().next().unwrap();
                            Ok((Some(operation_name.to_string()), operation))
                        }
                        None => Err(ValidationError::MultipleOperationsNoOperationName),
                        Some(operation_name) => {
                            let operation = operations.remove(&Name::new(&operation_name));

                            match operation {
                                None => {
                                    Err(ValidationError::MultipleOperationsUnmatchedOperationName(
                                        operation_name,
                                    ))
                                }
                                Some(operation) => Ok((Some(operation_name), operation)),
                            }
                        }
                    }
                }
            }
        }?;

        let operation_validator = OperationValidator::new(
            self.system,
            self.schema,
            operation_name,
            self.variables,
            document.fragments,
        );

        operation_validator.validate(raw_operation)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use async_graphql_parser::parse_query;

    // Macro to avoid insta-related issues with the test names when asserted through a helper function
    // It accepts arguments in the same way as that would be passed to the graphql endpoint
    macro_rules! assert_validation {
        ($query:expr, $operation_name:expr, $variables:expr) => {
            let system = create_test_system();
            let schema = Schema::new(&system);
            let variables = if $variables.is_empty() {
                None
            } else {
                Some(create_variables($variables))
            };

            let operation_name = if $operation_name.is_empty() {
                None
            } else {
                Some($operation_name.to_string())
            };

            let validator = DocumentValidator {
                system: &system,
                schema: &schema,
                operation_name,
                variables,
            };

            insta::assert_debug_snapshot!(validator.validate(create_query_document($query)));
        };
    }

    #[test]
    fn no_arguments_valid() {
        assert_validation!(
            r#"
                query {
                    concerts {
                        id
                        title
                        venue {
                            id
                            name
                        }
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn primitive_argument_valid() {
        assert_validation!(
            r#"
                query {
                    concert(id: 1) {
                        id
                        title
                        venue {
                            id
                            name
                        }
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn composite_argument_valid() {
        assert_validation!(
            r#"
                query {
                    concerts(where: {id: {eq: 1}}) {
                        id
                        title
                        venue {
                            id
                            name
                        }
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn composite_list_argument_valid() {
        assert_validation!(
            r#"
                query {
                    concerts(where: {and: [{id: {gt: 1}}, {id: {lt: 3}}]}) {
                        id
                        title
                        venue {
                            id
                            name
                        }
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn composite_order_by_valid_implicit_array() {
        assert_validation!(
            r#"
                query {
                    concerts(orderBy: {title: DESC}) {
                        id
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn composite_order_by_valid_explicit_array() {
        assert_validation!(
            r#"
                query {
                    concerts(orderBy: [{id: ASC}, {title: DESC}]) {
                        id
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn composite_order_by_invalid_overly_nested_array() {
        assert_validation!(
            r#"
                query {
                    concerts(orderBy: [[{id: ASC}, {title: DESC}]]) {
                        id
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn with_operation_name_valid() {
        assert_validation!(
            r#"
                query ConcertById {
                    concert(id: 1) {
                        id
                        title
                        venue {
                            id
                            name
                        }
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn stray_argument_invalid() {
        assert_validation!(
            r#"
                query {
                    concert(id: 1, foo: "bar") { # foo is not a valid argument
                        id
                        title
                        venue {
                            id
                            name
                        }
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn unspecified_required_argument_invalid() {
        assert_validation!(
            r#"
                query {
                    concert { # id argument is required here
                        id
                        title
                        venue {
                            id
                            name
                        }
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn variable_resolution_valid() {
        assert_validation!(
            r#"
                query($concert_id: Int!, $venue_id: Int!) {
                    concert(id: $concert_id) { # id argument is required here
                        id
                        title
                    }
                    venue(id: $venue_id) {
                        id
                        name
                    }
                }
            "#,
            "",
            r#"
            {
                "concert_id": 1,
                "venue_id": 2
            }"#
        );
    }

    #[test]
    fn variable_resolution_invalid() {
        assert_validation!(
            r#"
            query($concert_id: Int!, $venue_id: Int!) { # venue_id is not a specified in variables
                concert(id: $concert_id) { # id argument is required here
                    id
                    title
                }
                venue(id: $venue_id) {
                    id
                    name
                }
            }
        "#,
            "",
            r#"{ "concert_id": 2 }"#
        );
    }

    #[test]
    fn invalid_subfield() {
        assert_validation!(
            r#"
                query {
                    concert(id: 1) {
                        id
                        title
                        foobar
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn aliases_valid() {
        assert_validation!(
            r#"
                query {
                    firstConcert: concert(id: 1) {
                        id
                        headLine: title
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn multi_operations_valid() {
        let query = r#"
            query concert1 {
                concert(id: 1) {
                    id
                    headLine: title
                }
            }

            query concert2 {
                concert(id: 2) {
                    id
                    headLine: title
                }
            }
        "#;
        assert_validation!(query, "concert1", "");
        assert_validation!(query, "concert2", "");
    }

    #[test]
    fn multi_operations_no_operation_name_invalid() {
        assert_validation!(
            r#"
                query concert1 {
                    concert(id: 1) {
                        id
                        headLine: title
                    }
                }

                query concert2 {
                    concert(id: 2) {
                        id
                        headLine: title
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn multi_operations_mismatched_operation_name_invalid() {
        assert_validation!(
            r#"
                query concert1 {
                    concert(id: 1) {
                        id
                        headLine: title
                    }
                }

                query concert2 {
                    concert(id: 2) {
                        id
                        headLine: title
                    }
                }
            "#,
            "foo",
            ""
        );
    }

    #[test]
    fn basic_mutation() {
        assert_validation!(
            r#"
                mutation {
                    createConcert(data: {title: "Concert 1", venue: {id: 1}}) {
                        id
                        title
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn create_mutation_list_input() {
        assert_validation!(
            r#"
                mutation {
                    createConcerts(data: [{title: "Concert 1", venue: {id: 1}}, {title: "Concert 2", venue: {id: 2}}]) {
                        id
                        title
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn create_mutation_invalid_overly_nested_list_input() {
        assert_validation!(
            r#"
                mutation {
                    createConcerts(data: [[{title: "Concert 1", venue: {id: 1}}]]) {
                        id
                        title
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn service_mutation() {
        assert_validation!(
            r#"
                mutation {
                    processInfo(info: { level: 1, message: "allowed" })
                }
            "#,
            "",
            ""
        );
    }

    fn create_variables(variables: &str) -> Map<String, Value> {
        serde_json::from_str(variables).unwrap()
    }

    fn create_test_system() -> ModelSystem {
        let test_clay = r#"
            model Concert {
                id: Int = autoincrement() @pk
                title: String
                venue: Venue
            }

            model Venue {
                id: Int = autoincrement() @pk
                name: String
                concerts: Set<Concert>
            }

            @external("test.js")
            service TestService {
                input type Info {
                    level: Int
                    message: String
                }
            
                mutation processInfo(info: Info): Boolean
            }
        "#;
        payas_parser::build_system_from_str(test_clay, "test.clay".to_string()).unwrap()
    }

    fn create_query_document(query_str: &str) -> ExecutableDocument {
        parse_query(query_str).unwrap()
    }
}
