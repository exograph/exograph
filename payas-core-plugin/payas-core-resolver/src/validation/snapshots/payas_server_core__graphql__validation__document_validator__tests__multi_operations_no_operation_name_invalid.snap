---
source: payas-server-core/src/graphql/validation/document_validator.rs
expression: validator.validate(create_query_document(query))
---
Err(
    MultipleOperationsNoOperationName,
)
